### PR TITLE
Step and field conditionality

### PR DESCRIPTION
**Pull request by angusmcleod**

### CODE REVIEW

File: assets/javascripts/wizard/models/custom.js.es6 [hunk 1]
```diff
@@ -110,12 +94,45 @@ export function findCustomWizard(wizardId, params = {}) {
...
+        wizardJson.uncategorized_category_id
       );
     }
 
-    return CustomWizard.create(wizard);
+    return CustomWizard.create(wizardJson);
+  },
+});
+
+export function findCustomWizard(wizardId, params = {}) {
```
Comment (fzngagan) [inline assets/javascripts/wizard/models/custom.js.es6:105]: findCustomWizard -> fetchCustomWizard
The naming convention should suggest that the function is bring the data from the server. Thoughts?

File: assets/javascripts/wizard/models/custom.js.es6 [hunk 2]
```diff
@@ -110,12 +94,45 @@ export function findCustomWizard(wizardId, params = {}) {
...
+  }
+
+  return ajax({ url, cache: false, dataType: "json" }).then((result) => {
+    return CustomWizard.build(result);
   });
 }
 
+var _wizard_store;
+
+export function updateWizard(wizard) {
```
Comment (fzngagan) [inline assets/javascripts/wizard/models/custom.js.es6:130]: Similarly, the naming convention should suggest we're updating the local copy.

File: spec/requests/custom_wizard/steps_controller_spec.rb [hunk 3]
```diff
@@ -27,17 +39,121 @@
...
+    expect(response.status).to eq(200)
+
+    wizard_id = response.parsed_body['wizard']['id']
+    wizard = CustomWizard::Wizard.create(wizard_id, user)
+    group_name = wizard.submissions.last['action_9']
+    group = Group.find_by(name: group_name)
+    expect(group.full_name).to eq("My cool group")
+  end
+
+  it "detects the final step correctly" do
```
Comment (fzngagan) [inline spec/requests/custom_wizard/steps_controller_spec.rb:126]: We need to add a testcase to ensure whether a step can act as a final and non final step for different wizard paths.

**Comment by fzngagan** [comment]
@angusmcleod 
I've added a few points which could be improved.
Also, there are some conflicts which probably got introduced due to our recent code formatting changes. Need to get them resolved too. :)

File: assets/javascripts/wizard/models/custom.js.es6 (hunk 1 above)
Comment (angusmcleod) [reply assets/javascripts/wizard/models/custom.js.es6:105]: @fzngagan Maybe. This naming convention is bot pre-existing in this repo, and [copied from the core wizard](https://github.com/discourse/discourse/blob/master/app/assets/javascripts/wizard/models/wizard.js#L87) though, so I don't think we should change that in this PR.

File: assets/javascripts/wizard/models/custom.js.es6 (hunk 2 above)
Comment (angusmcleod) [reply assets/javascripts/wizard/models/custom.js.es6]: @fzngagan Agreed on this one :+1: Updated.

File: spec/requests/custom_wizard/steps_controller_spec.rb (hunk 3 above)
Comment (angusmcleod) [reply spec/requests/custom_wizard/steps_controller_spec.rb]: @fzngagan I've updated the controller and spec to add more error handling. After looking closer at the code, I don't think it makes sense to have a single test for that case, as it is a combination of other tests. The logic doesn't work without the force_final attribute. Remember that that has a role to play in both in the initial step serialization and the subsequent step submission. The controller is about the second part, the step submission. I'll look at adding more tests for the step serialization piece tomorrow.

**Comment by angusmcleod** [comment]
@fzngagan I've address the points and fixed the conflicts. There will be additional linting issues to fix probably, but I'll look at that again once we've merged the github ci actions, which we'll merge in here before merging this

**Comment by angusmcleod** [comment]
@fzngagan Ready for review :)

**Comment by fzngagan** [comment]
@angusmcleod 
Whoa!! Checks green. I'll check it briefly and merge.

**Pull request merged**